src/mbgl/style/expression/distance.cpp

```diff
@@ -0,0 +1,601 @@
+}
+
+double lineToLineDistance(const mapbox::geometry::line_string<double>& line1,
+                          IndexRange& range1,
+                          const mapbox::geometry::line_string<double>& line2,
+                          IndexRange& range2,
+                          mapbox::cheap_ruler::CheapRuler& ruler) {
+    double dist = std::numeric_limits<double>::infinity();
+    for (std::size_t i = range1.first; i < range1.second; ++i) {
+        const auto& p1 = line1[i];
+        const auto& p2 = line1[i + 1];
+        for (std::size_t j = range2.first; j < range2.second; ++j) {
+            const auto& q1 = line2[j];
+            const auto& q2 = line2[j + 1];
+            if (GeometryUtil<double>::segmentIntersectSegment(p1, p2, q1, q2)) return 0.;
+            auto dist1 = std::min(pointToLineDistance(p1, mapbox::geometry::line_string<double>{q1, q2}, ruler),
+                                  pointToLineDistance(p2, mapbox::geometry::line_string<double>{q1, q2}, ruler));
+            auto dist2 = std::min(pointToLineDistance(q1, mapbox::geometry::line_string<double>{p1, p2}, ruler),
+                                  pointToLineDistance(q2, mapbox::geometry::line_string<double>{p1, p2}, ruler));
+            dist = std::min(dist, std::min(dist1, dist2));
+        }
+    }
+    return dist;
+}
+
+double pointsToPointsDistance(const mapbox::geometry::multi_point<double>& points1,
+                              IndexRange& range1,
+                              const mapbox::geometry::multi_point<double>& points2,
+                              IndexRange& range2,
+                              mapbox::cheap_ruler::CheapRuler& ruler) {
+    double dist = std::numeric_limits<double>::infinity();
+    for (std::size_t i = range1.first; i <= range1.second; ++i) {
+        for (std::size_t j = range2.first; j <= range2.second; ++j) {
+            dist = std::min(dist, ruler.distance(points1[i], points2[j]));
+            if (dist == 0.) return dist;
+        }
+    }
+    return dist;
+}
+
+std::pair<mbgl::optional<IndexRange>, mbgl::optional<IndexRange>> splitRange(const IndexRange& range, bool isLine) {
+    auto size = getRangeSize(range);
+    if (isLine) {
+        if (size == 2) {
+            return std::make_pair(range, nullopt);
+        }
+        auto size1 = size / 2;
+        IndexRange range1(range.first, range.first + size1);
+        IndexRange range2(range.first + size1, range.second);
+        return std::make_pair(std::move(range1), std::move(range2));
+    } else {
+        if (size == 1) {
+            return std::make_pair(range, nullopt);
+        }
+        auto size1 = size / 2 - 1;
+        IndexRange range1(range.first, range.first + size1);
+        IndexRange range2(range.first + size1 + 1, range.second);
+        return std::make_pair(std::move(range1), std::move(range2));
+    }
+}
+
+// <distance, range1, range2>
+using DistPair = std::tuple<double, IndexRange, IndexRange>;
+class Comparator {
```

Comment:
could be free function (in a nameless namespace)

Reply:
I use this customer class as the comparator of priority_queue, I tried to use lambda, but it does not work with alias-declaration.